TestResultSummaryService/routes/getTabularData.js

```diff
@@ -0,0 +1,56 @@
+const { TestResultsDB, ObjectID } = require( '../Database' );
+
+function exceedDate(jdkDate) {
+    return function(element) {
+       return parseInt(element.aggregateInfo[0].benchmarkProduct.slice(-8)) <= parseInt(jdkDate);
+    }
+}
+
+
+module.exports = async ( req, res ) => {
+    const data = [];
+    const db = new TestResultsDB();
+    //console.log("Request received: ", req.query.jdkVersion, req.query.jvmType, req.query.jdkDate, req.query.sdkResource);
+    
+    const datas = [];
+    // TODO: Use available api to get build directly
+    const query = {buildName: {$regex: ".*_" + 'openjdk' + req.query.jdkVersion.substring(1) + "_" + req.query.jvmType + ".*perf_.*"}};
+    
+    // Get list of distinct platforms and benchmarks that match the JDK pipeline
+    const platforms = await db.distinct("buildName", query);
+    const benchmarks = await db.distinct("aggregateInfo.benchmarkName", query);
+    
+    // Extra check to return builds whose sdkResource field does not exist or is null.
+    let sdkResourceQuery;
+    if (req.query.sdkResource === "null") {
+    	sdkResourceQuery = null;
+    } else {
+    	sdkResourceQuery = req.query.sdkResource;
+    }
+    
+    for (benchmarkIndex in benchmarks) {
+        let benchmarkQuery;
+        // Return all entries that match the current benchmark and platform
+        for (item in platforms) {
+            benchmarkQuery = {$and: [{ buildName: { $regex : platforms[item] }, sdkResource: sdkResourceQuery}, { aggregateInfo: { $elemMatch: { benchmarkName: benchmarks[benchmarkIndex] }}}
+            ] };
+            const result = await db.getData(benchmarkQuery).toArray();
+
+            // Remove all entries whose build date exceeds the chosen date
+            const exceedFilter = result.filter(exceedDate(req.query.jdkDate));
+            // Setting the latest build date from the available dates
+            const latestDate = Math.max.apply(Math, exceedFilter.map(function(o) { return parseInt(o.aggregateInfo[0].benchmarkProduct.slice(-8)); }));
+            // Remove all runs that are not the latest
+            const dateFilter = exceedFilter.filter(entry => parseInt(entry.aggregateInfo[0].benchmarkProduct.slice(-8)) === latestDate);
+            const latest = Math.max.apply(Math, dateFilter.map(function(o) { return o.timestamp; }));
+            // Keep the latest build with the latest timestamp
+            const latestRun = dateFilter.find(function(o){ return o.timestamp == latest; })
+
+            if (latestRun !== undefined) {datas.push( latestRun );}
+        }
+    }
```

Comment:
> There is no need to use distinct twice.

Could you please clarify how can we get the distinct platforms, benchmark names and others such as Java version, type, SDK Resource (once we soft-code by querying from the database) without using `distinct`? Will check whether there's any other efficient way of doing it but haven't seen one yet. 

> We can query by the date and buildName and sort jdkBuildDateUnixTime in testData group on the buildName and benchmarkName.

I don't think we can query by a specific date. We want to get the latest JDK build that was built on or "before" the selected JDK date so we can't really query for one particular date. We use the `jdkBuildDateUnixTime` only when the latest JDK build that met the JDK date filter has multiple runs. So it could be the case that some JDK with a older built data could have a newer `jdkBuildDateUnixTime`. Currently, we're looking at the JDK built data in `benchmarkProduct`, something that we agreed to use for now but could be changed later.

Since we care about the latest run for each benchmark and platform,  I'm not sure how we can query in any other way. Sorry, I'm not getting your point. If that's the case, then could you please put the pseudo step-by-step code for it? We could also have a quick meeting if that's more convenient for you.